### PR TITLE
fix(dedup): unbounded GVN size-equivalence in hip-dedup-dps-inits (stacked on #416)

### PR DESCRIPTION
**Pull request by zpye**
﻿Stacked on #416 (base = `fix/whisper-cse-liveness`). Targets #416's branch, **not** `main`.

## Scope
Only the size-equivalence oracle of `hip-dedup-dps-inits` plus its regression test. (The doc / code-comment / `Pipelines.cpp`-comment edits that were briefly here have been removed — this PR is purely the fix.)

| File | Change |
|---|---|
| `lib/Dialect/Transforms/DedupDPSInits.cpp` | depth-limited size oracle -> unbounded GVN |
| `lib/Dialect/Transforms/CMakeLists.txt` | `+MLIRViewLikeInterface` (the alias-transitive liveness walk uses `ViewLikeOpInterface`) |
| `test/lit/Dialect/hip-dedup-dps-inits.mlir` | new: GVN cross-layer merge + size-mismatch negative |

## Problem in #416
`hip-dedup-dps-inits` merges two `tensor.empty` DPS inits only when (a) their buffers' live ranges are disjoint **and** (b) their sizes are provably equal. The size check (b) was the weak point, in two ways:

1. **Type-only equivalence over-merges.** Grouping same-typed empties without comparing their dynamic-extent operands merges two `tensor<?x...>` of *different* runtime size; the consumer then aliases a wrong/too-small buffer -> `wrap_slice` receives a garbage axis -> **runtime abort (`code 72`)** on the vision encoder.
2. **A depth-limited recursive structural compare is too weak.** The per-dim extent of a 4-D `reshape(-1)` is a deep product/`divsi` tree of `hip.readback_scalar`s. A fixed recursion depth exhausts on the tree and reports "not equal", so cross-layer activation buffers never merge and the activation pool blows up (~30 GB on the vision encoder).

## Fix
Replace the oracle with **global value numbering**. One bottom-up pass over the entry block (`BlockValueNumbering`) assigns every SSA value an equivalence-class id, equal iff produced by structurally-identical effect-free op chains over equal leaves. Two empties are the same size iff their dynamic-size operands share ids. This is exactly the equivalence a stock `-cse` folds to (CSE folds the shape arithmetic to a fixpoint first), but **unbounded-depth**, **without** CSE's unsafe empty-merge, and **with** the disjoint-liveness guard CSE lacks.

Enabler: pre-bufferize `hip.readback_scalar` is effect-free (`ReadbackScalarOp::getEffects` attaches the `Read` only once the operand is a `memref`), so per-layer extent arithmetic over a shared shape tensor unifies under GVN.

The disjoint-liveness half is unchanged: it walks the buffer's alias set transitively through DPS tied-results and `ViewLikeOpInterface` ops, so a buffer kept live by a later view/insert read is not under-counted.

## Verification (this build)
- **LIT:** full `MorphizenMLIRLitTests` suite passes; the new `hip-dedup-dps-inits.mlir` (positive GVN merge + negative size-mismatch) passes against the freshly built `hip-mlir-opt`.
- **Vision encoder** (the `code 72` reproducer), `HIPDNN_EP_STRICT=1`, model cache cleared: `[DedupDPSInits] merged 855 tensor.empty ops`; runs end-to-end, **no `code 72`** (~30 GB -> ~1.2 GB activation pool, == `-cse`-on baseline).
- **Llama-1B** (cache cleared): `test_ort_dynamic_vs_fixed` KV **cosine = 1.0** (bit-identical); `test_ort_per_step_logits` prefill + 9 decode steps **cosine ~0.9999, top1 match** -> no accuracy regression.
- **Whisper fp16 encoder** cosine **0.999984** (the CSE-clobber guard #416 added is preserved).

## Test plan
- [ ] `ctest --test-dir install/build -C RelWithDebInfo -R MorphizenMLIRLitTests`
- [ ] Vision encoder compiles + runs with `HIPDNN_EP_STRICT=1` (no `code 72`), activation pool ~1.2 GB
- [ ] Llama-1B accuracy unchanged (dynamic-vs-fixed KV cosine 1.0; per-step logits cosine ~0.9999)
- [ ] Whisper fp16 encoder cosine ~0.99998

### CODE REVIEW

**Comment by amd-mingw** [comment]
http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/1085/
http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/1087/

**Comment by amd-yanjunz** [comment]
ROCm OGA CI finished with result: UNSTABLE.

- PR: ROCm/onnx-hipdnn-ep#419
- Source comment: [modelzoo=LLM_ROCm_Whisper models=all](https://github.com/ROCm/onnx-hipdnn-ep/pull/419#issuecomment-4810710893)
- TEST_MODE: both
- MODELZOO: LLM_ROCm_Whisper
- TEST_MODELS: (all)
- Jenkins build: http://xsjvitisaijenkins:8080/job/daily_hipdnn_ep_oga/1088/
- NTS report: http://xsjnts.xilinx.com/ROCm/hipdnn-detail?database=development&collection=daily_hipdnn_ep_oga&id=6a3e9ba97314e9a62bda6cb7
